solidity/test/token-distribution/TokenDistributorTest.js

```diff
@@ -0,0 +1,651 @@
+      )
+
+      const timestamp = toBN(
+        (await web3.eth.getBlock(receipt.blockNumber)).timestamp
+      )
+
+      expect(
+        await tokenDistributor.unclaimedUnlockTimestamp(),
+        "invalid unclaimed unlock timestamp"
+      ).to.eq.BN(timestamp.add(unclaimedUnlockDuration))
+    })
+
+    it("doesn't set unclaimed tokens unlock timestamp when unclaimed duration is not provided", async function () {
+      await tokenDistributor.allocate(
+        testData.merkle.merkleRoot,
+        testData.merkle.tokenTotal,
+        0,
+        { from: owner }
+      )
+
+      expect(
+        await tokenDistributor.unclaimedUnlockTimestamp(),
+        "invalid unclaimed unlock timestamp"
+      ).to.eq.BN(0)
+    })
+
+    it("reverts on merkle root overwrite", async function () {
+      await tokenDistributor.allocate(
+        testData.merkle.merkleRoot,
+        testData.merkle.tokenTotal,
+        unclaimedUnlockDuration,
+        { from: owner }
+      )
+
+      await expectRevert(
+        tokenDistributor.allocate(
+          "0x1234567890",
+          testData.merkle.tokenTotal,
+          unclaimedUnlockDuration,
+          {
+            from: owner,
+          }
+        ),
+        "tokens were already allocated"
+      )
+    })
+
+    it("reverts on empty merkle root", async function () {
+      await expectRevert(
+        tokenDistributor.allocate(
+          [],
+          testData.merkle.tokenTotal,
+          unclaimedUnlockDuration,
+          {
+            from: owner,
+          }
+        ),
+        "merkle root cannot be empty"
+      )
+    })
+
+    it("reverts on zero amount", async function () {
+      await expectRevert(
+        tokenDistributor.allocate(
+          testData.merkle.merkleRoot,
+          0,
+          unclaimedUnlockDuration,
+          {
+            from: owner,
+          }
+        ),
+        "amount has to be greater than zero"
+      )
+    })
+
+    it("reverts on token transfer failure", async function () {
+      await expectRevert(
+        tokenDistributor.allocate(
+          testData.merkle.merkleRoot,
+          testData.merkle.tokenTotal + 1,
+          unclaimedUnlockDuration,
+          {
+            from: owner,
+          }
+        ),
+        "SafeERC20: low-level call failed"
+      )
+    })
+
+    it("reverts when called by non-owner", async function () {
+      await expectRevert(
+        tokenDistributor.allocate(
+          testData.merkle.merkleRoot,
+          testData.merkle.tokenTotal,
+          unclaimedUnlockDuration,
+          {
+            from: thirdParty,
+          }
+        ),
+        "Ownable: caller is not the owner"
+      )
+    })
+  })
+
+  describe("claim", async () => {
+    before(async () => {
+      await tokenDistributor.allocate(
+        testData.merkle.merkleRoot,
+        testData.merkle.tokenTotal,
+        unclaimedUnlockDuration,
+        { from: owner }
+      )
+    })
+
+    after(freshDeployment)
```

Comment:
This is confusing. I'd rather drop it and changed `before` into `beforeEach`. WDYT?

Reply:
Yeah, `beforeEach` will do the work!